infra/Pos/Communication/Relay/Types.hs

```diff
@@ -44,10 +45,17 @@ instance Buildable PropagationMsg where
     build (DataOnlyPM conts) =
         Buildable.build conts
 
--- | Queue of InvMsges which should be propagated.
-type RelayPropagationQueue = TBQueue (Microsecond, PropagationMsg)
+data RelayContext m = RelayContext
```

Comment:
Regarding `AllOtherDataToImplementRelayMethods`: the difficulty is that prior to this change the relay context would use concrete things (an `STM` channel), but now we want to use something abstract, so that implementations in time-warp can handle the details. It's the difference between offering a `TChan t` versus an interface with `enqueue :: t -> IO ()` and `dequeue :: IO t`. The former doesn't need to mention the monad, but the latter must.

We want to be able to use a relayer abstraction which, at enqueue time, can use effects from the application monad (call it `m`). This will be one of those `-Mode` monads like `RealMode`. That's to say, to enqueue a message for relay is now potentially more complicated than a simple `writeTChan`; it may, for instance, use some `m`-term to come up with the next slot leader and use that information to prioritize it. So our relay context must not, implicitly (as in master) or explicitly (as in this patch) pin us to `IO` or even just to `MonadIO` for enqueue; it has to allow for the full range of effects offered by the application monad.

I'll push a patch in the morning (~10 hours from now) which I believe improves the situation regarding node resources. The `RelayContext` still has the type parameter `m`, and it's picked out by the `NodeResources ssc m` type. However, it's *not* the same as the monad in which the resources are created! Rather than `m (NodeResources ssc m)` for allocation, it's `n (NodeResources ssc m)` where `m` has some constraints. This eliminates the need for various `hoist-`s; unification does that work instead.